File: basic_dsa_twenty_two.py
"""
Basic DSA Problems - Two More Fundamental Questions
Simple and essential DSA problems for beginners
"""


# Problem 1: Check if a Number is Disarium Number
# A Disarium number is a number where the sum of its digits raised to their respective positions equals the number itself
# Example: 135 = 1^1 + 3^2 + 5^3 = 1 + 9 + 125 = 135

def count_digits(n):
    """
    Count number of digits in a number
    Time Complexity: O(log n)
    Space Complexity: O(1)
    """
    if n == 0:
        return 1
    count = 0
    num = abs(n)
    while num > 0:
        count += 1
        num //= 10
    return count


def is_disarium_number(n):
    """
    Check if a number is Disarium number
    Time Complexity: O(d * log n) where d is number of digits
    Space Complexity: O(1)
    """
    if n <= 0:
        return False
    
    original = n
    num_digits = count_digits(n)
    sum_powers = 0
    
    # Process digits from right to left
    position = num_digits
    while n > 0:
        digit = n % 10
        sum_powers += digit ** position
        position -= 1
        n //= 10
    
    return sum_powers == original


def get_disarium_numbers_in_range(start, end):
    """
    Find all Disarium numbers in a given range
    Time Complexity: O((end - start) * d * log n)
    Space Complexity: O(k) where k is the number of Disarium numbers
    """
    disarium_numbers = []
    
    for num in range(start, end + 1):
        if is_disarium_number(num):
            disarium_numbers.append(num)
    
    return disarium_numbers


def get_digit_powers_sum(n):
    """
    Get the sum of digits raised to their positions
    Time Complexity: O(d * log n) where d is number of digits
    Space Complexity: O(1)
    """
    if n <= 0:
        return 0
    
    num_digits = count_digits(n)
    sum_powers = 0
    position = num_digits
    
    while n > 0:
        digit = n % 10
        sum_powers += digit ** position
        position -= 1
        n //= 10
    
    return sum_powers


# Problem 2: Convert Octal to Decimal
# Convert an octal (base 8) number to its decimal representation

def octal_to_decimal_iterative(octal_str):
    """
    Convert octal string to decimal using iteration
    Time Complexity: O(n) where n is length of octal string
    Space Complexity: O(1)
    """
    if not octal_str or octal_str == "0":
        return 0
    
    is_negative = octal_str[0] == '-'
    start = 1 if is_negative else 0
    
    decimal = 0
    power = 0
    
    # Process from right to left
    for i in range(len(octal_str) - 1, start - 1, -1):
        digit = int(octal_str[i])
        if digit < 0 or digit > 7:
            raise ValueError(f"Invalid octal digit: {digit}")
        decimal += digit * (8 ** power)
        power += 1
    
    return -decimal if is_negative else decimal


def octal_to_decimal_recursive(octal_str):
    """
    Convert octal string to decimal using recursion
    Time Complexity: O(n) where n is length of octal string
    Space Complexity: O(n) due to recursion stack
    """
    def helper(octal, index, power):
        if index < 0 or (index == 0 and octal[0] == '-'):
            return 0
        digit = int(octal[index])
        if digit < 0 or digit > 7:
            raise ValueError(f"Invalid octal digit: {digit}")
        return digit * (8 ** power) + helper(octal, index - 1, power + 1)
    
    if not octal_str or octal_str == "0":
        return 0
    
    is_negative = octal_str[0] == '-'
    start = len(octal_str) - 1
    
    result = helper(octal_str, start, 0)
    return -result if is_negative else result


def octal_to_decimal_builtin(octal_str):
    """
    Convert octal string to decimal using built-in function
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if not octal_str:
        return 0
    
    is_negative = octal_str[0] == '-'
    octal = octal_str[1:] if is_negative else octal_str
    
    decimal = int(octal, 8)
    return -decimal if is_negative else decimal


def octal_array_to_decimal(octal_array):
    """
    Convert octal array to decimal
    Time Complexity: O(n) where n is length of array
    Space Complexity: O(1)
    """
    if not octal_array:
        return 0
    
    decimal = 0
    power = len(octal_array) - 1
    
    for digit in octal_array:
        if digit < 0 or digit > 7:
            raise ValueError(f"Invalid octal digit: {digit}")
        decimal += digit * (8 ** power)
        power -= 1
    
    return decimal


# Test cases
if __name__ == "__main__":
    print("=" * 60)
    print("Problem 1: Check if a Number is Disarium Number")
    print("=" * 60)
    
    # Test Case 1 - Disarium number 135
    num1 = 135
    result1 = is_disarium_number(num1)
    sum1 = get_digit_powers_sum(num1)
    print(f"Number: {num1}")
    print(f"Sum of digits raised to positions: {sum1}")
    print(f"Is Disarium number: {result1}")
    
    # Test Case 2 - Disarium number 89
    num2 = 89
    result2 = is_disarium_number(num2)
    sum2 = get_digit_powers_sum(num2)
    print(f"\nNumber: {num2}")
    print(f"Sum of digits raised to positions: {sum2}")
    print(f"Is Disarium number: {result2}")
    
    # Test Case 3 - Not Disarium number
    num3 = 123
    result3 = is_disarium_number(num3)
    sum3 = get_digit_powers_sum(num3)
    print(f"\nNumber: {num3}")
    print(f"Sum of digits raised to positions: {sum3}")
    print(f"Is Disarium number: {result3}")
    
    # Test Case 4 - Single digit Disarium
    num4 = 1
    result4 = is_disarium_number(num4)
    print(f"\nNumber: {num4}")
    print(f"Is Disarium number: {result4}")
    
    # Test Case 5 - Find Disarium numbers in range
    start = 1
    end = 200
    disarium_nums = get_disarium_numbers_in_range(start, end)
    print(f"\nDisarium numbers in range [{start}, {end}]:")
    print(disarium_nums)
    print(f"Count: {len(disarium_nums)}")
    
    print("\n" + "=" * 60)
    print("Problem 2: Convert Octal to Decimal")
    print("=" * 60)
    
    # Test Case 1 - Basic octal
    octal1 = "123"
    result1_iter = octal_to_decimal_iterative(octal1)
    result1_rec = octal_to_decimal_recursive(octal1)
    result1_builtin = octal_to_decimal_builtin(octal1)
    print(f"Octal: {octal1}")
    print(f"Decimal (iterative): {result1_iter}")
    print(f"Decimal (recursive): {result1_rec}")
    print(f"Decimal (built-in): {result1_builtin}")
    
    # Test Case 2 - Zero
    octal2 = "0"
    result2 = octal_to_decimal_iterative(octal2)
    print(f"\nOctal: {octal2}")
    print(f"Decimal: {result2}")
    
    # Test Case 3 - Large octal
    octal3 = "377"
    result3 = octal_to_decimal_iterative(octal3)
    print(f"\nOctal: {octal3}")
    print(f"Decimal: {result3}")
    
    # Test Case 4 - Octal array
    octal4 = [1, 2, 7]
    result4 = octal_array_to_decimal(octal4)
    print(f"\nOctal (array): {octal4}")
    print(f"Decimal: {result4}")
    
    # Test Case 5 - Single digit
    octal5 = "7"
    result5 = octal_to_decimal_iterative(octal5)
    print(f"\nOctal: {octal5}")
    print(f"Decimal: {result5}")

File: basic_dsa_twenty_two.java
/**
 * Basic DSA Problems - Two More Fundamental Questions
 * Simple and essential DSA problems for beginners
 */
import java.util.*;

public class basic_dsa_twenty_two {
    
    // Problem 1: Check if a Number is Disarium Number
    // A Disarium number is a number where the sum of its digits raised to their respective positions equals the number itself
    // Example: 135 = 1^1 + 3^2 + 5^3 = 1 + 9 + 125 = 135
    
    /**
     * Count number of digits in a number
     * Time Complexity: O(log n)
     * Space Complexity: O(1)
     */
    public static int countDigits(int n) {
        if (n == 0) {
            return 1;
        }
        int count = 0;
        int num = Math.abs(n);
        while (num > 0) {
            count++;
            num /= 10;
        }
        return count;
    }
    
    /**
     * Check if a number is Disarium number
     * Time Complexity: O(d * log n) where d is number of digits
     * Space Complexity: O(1)
     */
    public static boolean isDisariumNumber(int n) {
        if (n <= 0) {
            return false;
        }
        
        int original = n;
        int numDigits = countDigits(n);
        long sumPowers = 0;
        
        // Process digits from right to left
        int position = numDigits;
        while (n > 0) {
            int digit = n % 10;
            sumPowers += (long) Math.pow(digit, position);
            position--;
            n /= 10;
        }
        
        return sumPowers == original;
    }
    
    /**
     * Find all Disarium numbers in a given range
     * Time Complexity: O((end - start) * d * log n)
     * Space Complexity: O(k) where k is the number of Disarium numbers
     */
    public static ArrayList<Integer> getDisariumNumbersInRange(int start, int end) {
        ArrayList<Integer> disariumNumbers = new ArrayList<>();
        
        for (int num = start; num <= end; num++) {
            if (isDisariumNumber(num)) {
                disariumNumbers.add(num);
            }
        }
        
        return disariumNumbers;
    }
    
    /**
     * Get the sum of digits raised to their positions
     * Time Complexity: O(d * log n) where d is number of digits
     * Space Complexity: O(1)
     */
    public static long getDigitPowersSum(int n) {
        if (n <= 0) {
            return 0;
        }
        
        int numDigits = countDigits(n);
        long sumPowers = 0;
        int position = numDigits;
        int temp = n;
        
        while (temp > 0) {
            int digit = temp % 10;
            sumPowers += (long) Math.pow(digit, position);
            position--;
            temp /= 10;
        }
        
        return sumPowers;
    }
    
    // Problem 2: Convert Octal to Decimal
    // Convert an octal (base 8) number to its decimal representation
    
    /**
     * Convert octal string to decimal using iteration
     * Time Complexity: O(n) where n is length of octal string
     * Space Complexity: O(1)
     */
    public static int octalToDecimalIterative(String octalStr) {
        if (octalStr == null || octalStr.isEmpty() || octalStr.equals("0")) {
            return 0;
        }
        
        boolean isNegative = octalStr.charAt(0) == '-';
        int start = isNegative ? 1 : 0;
        
        int decimal = 0;
        int power = 0;
        
        // Process from right to left
        for (int i = octalStr.length() - 1; i >= start; i--) {
            int digit = Character.getNumericValue(octalStr.charAt(i));
            if (digit < 0 || digit > 7) {
                throw new IllegalArgumentException("Invalid octal digit: " + digit);
            }
            decimal += digit * Math.pow(8, power);
            power++;
        }
        
        return isNegative ? -decimal : decimal;
    }
    
    /**
     * Convert octal string to decimal using recursion
     * Time Complexity: O(n) where n is length of octal string
     * Space Complexity: O(n) due to recursion stack
     */
    public static int octalToDecimalRecursive(String octalStr) {
        if (octalStr == null || octalStr.isEmpty() || octalStr.equals("0")) {
            return 0;
        }
        
        boolean isNegative = octalStr.charAt(0) == '-';
        String octal = isNegative ? octalStr.substring(1) : octalStr;
        
        int result = helper(octal, octal.length() - 1, 0);
        return isNegative ? -result : result;
    }
    
    private static int helper(String octal, int index, int power) {
        if (index < 0) {
            return 0;
        }
        int digit = Character.getNumericValue(octal.charAt(index));
        if (digit < 0 || digit > 7) {
            throw new IllegalArgumentException("Invalid octal digit: " + digit);
        }
        return digit * (int) Math.pow(8, power) + helper(octal, index - 1, power + 1);
    }
    
    /**
     * Convert octal string to decimal using built-in function
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    public static int octalToDecimalBuiltin(String octalStr) {
        if (octalStr == null || octalStr.isEmpty()) {
            return 0;
        }
        
        boolean isNegative = octalStr.charAt(0) == '-';
        String octal = isNegative ? octalStr.substring(1) : octalStr;
        
        int decimal = Integer.parseInt(octal, 8);
        return isNegative ? -decimal : decimal;
    }
    
    /**
     * Convert octal array to decimal
     * Time Complexity: O(n) where n is length of array
     * Space Complexity: O(1)
     */
    public static int octalArrayToDecimal(int[] octalArray) {
        if (octalArray == null || octalArray.length == 0) {
            return 0;
        }
        
        int decimal = 0;
        int power = octalArray.length - 1;
        
        for (int digit : octalArray) {
            if (digit < 0 || digit > 7) {
                throw new IllegalArgumentException("Invalid octal digit: " + digit);
            }
            decimal += digit * Math.pow(8, power);
            power--;
        }
        
        return decimal;
    }
    
    // Test cases
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem 1: Check if a Number is Disarium Number");
        System.out.println("=".repeat(60));
        
        // Test Case 1 - Disarium number 135
        int num1 = 135;
        boolean result1 = isDisariumNumber(num1);
        long sum1 = getDigitPowersSum(num1);
        System.out.println("Number: " + num1);
        System.out.println("Sum of digits raised to positions: " + sum1);
        System.out.println("Is Disarium number: " + result1);
        
        // Test Case 2 - Disarium number 89
        int num2 = 89;
        boolean result2 = isDisariumNumber(num2);
        long sum2 = getDigitPowersSum(num2);
        System.out.println("\nNumber: " + num2);
        System.out.println("Sum of digits raised to positions: " + sum2);
        System.out.println("Is Disarium number: " + result2);
        
        // Test Case 3 - Not Disarium number
        int num3 = 123;
        boolean result3 = isDisariumNumber(num3);
        long sum3 = getDigitPowersSum(num3);
        System.out.println("\nNumber: " + num3);
        System.out.println("Sum of digits raised to positions: " + sum3);
        System.out.println("Is Disarium number: " + result3);
        
        // Test Case 4 - Single digit Disarium
        int num4 = 1;
        boolean result4 = isDisariumNumber(num4);
        System.out.println("\nNumber: " + num4);
        System.out.println("Is Disarium number: " + result4);
        
        // Test Case 5 - Find Disarium numbers in range
        int start = 1;
        int end = 200;
        ArrayList<Integer> disariumNums = getDisariumNumbersInRange(start, end);
        System.out.println("\nDisarium numbers in range [" + start + ", " + end + "]:");
        System.out.println(disariumNums);
        System.out.println("Count: " + disariumNums.size());
        
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Problem 2: Convert Octal to Decimal");
        System.out.println("=".repeat(60));
        
        // Test Case 1 - Basic octal
        String octal1 = "123";
        int result1_dec = octalToDecimalIterative(octal1);
        int result1_rec = octalToDecimalRecursive(octal1);
        int result1_builtin = octalToDecimalBuiltin(octal1);
        System.out.println("Octal: " + octal1);
        System.out.println("Decimal (iterative): " + result1_dec);
        System.out.println("Decimal (recursive): " + result1_rec);
        System.out.println("Decimal (built-in): " + result1_builtin);
        
        // Test Case 2 - Zero
        String octal2 = "0";
        int result2_dec = octalToDecimalIterative(octal2);
        System.out.println("\nOctal: " + octal2);
        System.out.println("Decimal: " + result2_dec);
        
        // Test Case 3 - Large octal
        String octal3 = "377";
        int result3_dec = octalToDecimalIterative(octal3);
        System.out.println("\nOctal: " + octal3);
        System.out.println("Decimal: " + result3_dec);
        
        // Test Case 4 - Octal array
        int[] octal4 = {1, 2, 7};
        int result4_dec = octalArrayToDecimal(octal4);
        System.out.println("\nOctal (array): " + Arrays.toString(octal4));
        System.out.println("Decimal: " + result4_dec);
        
        // Test Case 5 - Single digit
        String octal5 = "7";
        int result5_dec = octalToDecimalIterative(octal5);
        System.out.println("\nOctal: " + octal5);
        System.out.println("Decimal: " + result5_dec);
    }
}
